rfc/1216-resolve-cloud-tokens.md

```diff
@@ -0,0 +1,845 @@
+```ts
+import * as aws from 'aws-cdk-lib';
+import * as k8s from 'cdk8s';
+import * as kplus from 'cdk8s-plus-26';
+
+import { AwsCdkTokenResolver } from '@cdk8s/aws-cdk-token-resolver';
+
+const awsApp = new aws.App();
+const stack = new aws.Stack(awsApp, 'Stack');
+
+const resolver = new AwsCdkTokenResolver(stack);
+
+const k8sApp = new k8s.App();
+const manifest = new k8s.Chart(k8sApp, 'Manifest', { resolver });
+
+const bucket = new aws.aws_s3.Bucket(stack, 'Bucket');
+
+new kplus.CronJob(manifest, 'CronJob', {
+  schedule: k8s.Cron.daily(),
+  containers: [{
+    image: 'job',
+    envVariables: {
+      // passing the bucket name via an env variable
+      BUCKET_NAME: kplus.EnvValue.fromValue(bucket.bucketName),
+    }
+ }]
+});
+
+awsApp.synth();
+k8sApp.synth();
+```
+
+Notice we create two applications: one for our cdk8s constructs, and one for our AWS CDK constructs.
+Both are defined and synthesized in the same file, but can be separated as needed.
+Since your Kubernetes resources now depend on AWS CDK deploy tokens, you'll first need to run `cdk deploy`, 
+and only then `cdk8s synth`.
+
+> Otherwise, the Kubernetes manifests will contain a string representation of the tokens (e.g `${Token[TOKEN.25]}`), 
+> instead of the concrete values. To learn more AWS CDK tokens, see [here](https://docs.aws.amazon.com/cdk/v2/guide/tokens.html).
+
+The implementation of the resolver utilizes [Cloud Control](https://docs.aws.amazon.com/cloudcontrolapi/latest/userguide/what-is-cloudcontrolapi.html)
+to fetch the runtime attributes of a resource. However, not all CloudFormation resources are currently supported by Cloud Control.
+
+> For a list of supported resources, see [here](https://docs.aws.amazon.com/cloudcontrolapi/latest/userguide/supported-resources.html).
+
+In case you need an attribute for such a resource, you can add a `CfnOutput` resource to your stack, and use the resolver to 
+explicitly fetch its value from the deployed stack. For example:
+
+```ts
+...
+
+// create the output
+const bucketNameOutput = new aws.CfnOutput(this, 'BucketName', {
+  value: bucket.bucketName,
+});
+
+const bucketName = resolver.fetchOutput(bucketNameOutput);
+
+new kplus.CronJob(manifest, 'CronJob', {
+  schedule: k8s.Cron.daily(),
+  containers: [{
+    image: 'job',
+    envVariables: {
+      // passing the output value
+      BUCKET_NAME: kplus.EnvValue.fromValue(bucketName),
+    }
+ }]
+});
+
+...
+```
+
+#### CDK For Terraform
+
+In this example, we create an S3 `Bucket` with the CDKTF, and pass its (deploy time generated) name 
+as an environment variable to a Kubernetes `CronJob` resource.
+
+```ts
+import * as tf from "cdktf";
+import * as aws from "@cdktf/provider-aws";
+import * as k8s from 'cdk8s';
+import * as kplus from 'cdk8s-plus-26';
+
+import { CdkTfTokenResolver } from '@cdk8s/cdktf-token-resolver';
+
+const awsApp = new tf.App();
+const stack = new tf.TerraformStack(awsApp, 'Stack');
+
+const resolver = new CdkTfTokenResolver(stack);
```

Comment:
How is the resolver checking the provenance of tokens? What if I want to use 2 resolvers at the same time (e.g., use values from AWS CDK and CDKTF both in my `Chart`?)

Reply:
Each resolver uses facilities provided by the CDK framework to check (i.e `Token.isUnresolved`). 

Currently the implementation only allows passing one resolver to the cdk8s app, so if you actually have values from both CDKTF and AWS CDK, e.g:

```ts
new ApiObject(this, 'obj', {
  spec: {
    prop1: <token-from-aws-cdk>,
    prop2: <token-from-cdktf>
  }
});
```

You would have to implement your own resolver that delegates to the specific ones. I suppose we could vend such a resolver ourselves if we see it is needed. But I doubt there are many users operating both AWS CDK and CDKTF...

Another interesting use case is:

```ts
new ApiObject(this, 'obj', {
  spec: {
    prop1: <token-from-aws-cdk> + <token-from-cdktf>,
  }
});
```

I haven't considered this, not even sure this is possible...but again I doubt this is an actual use case.
Interested in your thoughts here though.

Reply:
Given the resolver API we currently have, which I really like, I think we could trivially allow the first case by allowing this:

```ts
const k8sApp = new k8s.App({ 
  resolvers: [new AwsCdkResolver(), new CdkTfResolver()],
});
```

Iterating over all resolvers, and stopping as soon as one of them calls `context.replaceValue()`.

---------

The second case, where we have `token1 + token2` can only work if we can peel apart the separate tokens, and that can only work if the token formatting is standardized, which it currently is not.

So *shrug*, might be useful but can't be in scope for this.

Reply:
Yeah you're right about the first case - I'll make the change. Regarding the second one, if we see a demand for it, it might be a good forcing function to standardize tokens across the board and put them in `constructs` with a consistent formatting.